### PR TITLE
JDK-8289249: Add methods to Elements for record constructors

### PR DESCRIPTION
**Pull request by jddarcy**
Please review the addition of two predicates to better support records in the javax.lang.model API.

Please also review the CSR https://bugs.openjdk.org/browse/JDK-8291594

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8289249](https://bugs.openjdk.org/browse/JDK-8289249): Add methods to Elements for record constructors
 * [JDK-8291594](https://bugs.openjdk.org/browse/JDK-8291594): Add methods to Elements for record constructors (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9694/head:pull/9694` \
`$ git checkout pull/9694`

Update a local copy of the PR: \
`$ git checkout pull/9694` \
`$ git pull https://git.openjdk.org/jdk pull/9694/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9694`

View PR using the GUI difftool: \
`$ git pr show -t 9694`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9694.diff">https://git.openjdk.org/jdk/pull/9694.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/processing/model/util/elements/TestRecordPredicates.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+     * Annotation the class, not just constructor, since many of the
+     * constructors that are of interest are implicitly declared.
+     */
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface ExpectedPredicates {
+        boolean isCompact()   default false;
+        boolean isCanonical() default false;
+    }
+
+    @ExpectedPredicates(isCompact=true, isCanonical=true)
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/model/util/elements/TestRecordPredicates.java:110]: side: this is uncovering a bug in the compiler, isCanonical should be false in this case. I have filed JDK-8291914 to address this issue

**Review by vicente-romero-oracle — APPROVED** [review]
looks good to me

**Review by jonathan-gibbons — COMMENTED** [review]
Generally, when we specify an `@implNote` for a default method that is a "do-nothing-helpful" method, we should find a place or way to document the useful implementation of the method.

**Comment by jddarcy** [comment]
> Generally, when we specify an `@implNote` for a default method that is a "do-nothing-helpful" method, we should find a place or way to document the useful implementation of the method.

I think we could look to improve the docs on that point for future work.

File: test/langtools/tools/javac/processing/model/util/elements/TestRecordPredicates.java (hunk 1 above)
Comment (jddarcy) [reply test/langtools/tools/javac/processing/model/util/elements/TestRecordPredicates.java:110]: Part of the regression tests for  JDK-8291914 already written then :-)

Comment (vicente-romero-oracle) [reply test/langtools/tools/javac/processing/model/util/elements/TestRecordPredicates.java:110]: :)

**Pull request closed without merging**